src/orion/executor/ray_backend.py

```diff
@@ -0,0 +1,134 @@
+import logging
+import traceback
+
+from orion.executor.base import (
+    AsyncException,
+    AsyncResult,
+    BaseExecutor,
+    ExecutorClosed,
+    Future,
+)
+
+try:
+    import ray
+
+    HAS_RAY = True
+except ImportError:
+    HAS_RAY = False
+
+logger = logging.getLogger(__name__)
+
+
+class _Future(Future):
+    def __init__(self, future):
+        self.future = future
+        self.exception = None
+
+    def get(self, timeout=None):
+        if self.exception:
+            raise self.exception
+        try:
+            return ray.get(self.future, timeout=timeout)
+        except ray.exceptions.GetTimeoutError as e:
+            raise TimeoutError() from e
+
+    def wait(self, timeout=None):
+        try:
+            ray.get(self.future, timeout=timeout)
+        except ray.exceptions.GetTimeoutError:
+            pass
+        except Exception as e:
+            self.exception = e
+
+    def ready(self):
+        obj_ready = ray.wait([self.future])
+        return len(obj_ready[0]) == 1
+
+    def successful(self):
+        # Python 3.6 raise assertion error
+        if not self.ready():
+            raise ValueError()
+
+        return self.future.successful()
+
+
+class Ray(BaseExecutor):
+    def __init__(
+        self,
+        n_workers=-1,
+        runtime_env=None,
+        **config,
+    ):
+        super().__init__(n_workers=n_workers)
+        self.initialized = False
+        if not HAS_RAY:
+            raise ImportError("Ray must be installed to use Ray executor.")
+        self.config = config
+        if not ray.is_initialized():
+            if runtime_env is None:
+                ray.init()
+            else:
+                ray.init(
+                    runtime_env={
+                        "working_dir": "/Users/simonolivier/DevOrion/orion/tests/unittests/executor"
+                    }
+                )
+            self.initialized = True
+            print("Ray was initiated with runtime_env : ", runtime_env)
+
+    def __getstate__(self):
+        return super().__getstate__()
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+
+    def close(self):
+        if self.initialized:
+            self.initialized = False
+            ray.shutdown()
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
+        return self
+
+    def submit(self, function, *args, **kwargs):
+        try:
+            if ray.is_initialized():
+                remote_g = ray.remote(function)
+                return _Future(remote_g.remote(*args, **kwargs))
+            raise ExecutorClosed()
+        except Exception as e:
+            if str(e).startswith(
```

Comment:
Why catch the `ExecutorClosed()` previously raised and reraise it? Why catch it at all?

Reply:
I first went with the try - catch inspired by the Dask backend. But with ray, it wasnt calling an ExecutorClosed error, which was required in the tests. So I manually checked if ray is initiated, but left the try - catch. Should I remove it altogether ? 
The function would then only contain : 

`if not ray.is_initialized():
     raise ExecutorClosed()
            
  remote_g = ray.remote(function)         
  return _Future(remote_g.remote(*args, **kwargs))`

Reply:
Yeah there is no need for a try-catch here.